### PR TITLE
ISPN-6994 Explicit Subject impersonation

### PR DESCRIPTION
**Pull request by tristantarrant**
- Overcomes the need to keep the Subject in a ThreadLocal and avoids the use of PrivilegedActions in user code

https://issues.jboss.org/browse/ISPN-6994

### CODE REVIEW

**Comment by tristantarrant** [comment]
Still need to properly handle query permissions

**Comment by tristantarrant** [comment]
Failures are unrelated

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder10.java [hunk 1]
```diff
@@ -12,6 +12,8 @@
 import java.util.Map;
 import java.util.Optional;
 
+import javax.security.auth.Subject;
```
Comment (galderz) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder10.java]: Needed?

**Comment by tristantarrant** [comment]
Hey @galderz, I removed the unneeded import

**Comment by tristantarrant** [comment]
I have renamed as() to withSubject() which follows withFlags()

**Comment by ryanemerson** [comment]
@tristantarrant The `NoSuchMethodException`'s in `org.infinispan.security.*` appear to be related to your latest changes

**Comment by tristantarrant** [comment]
I've fixed the SecureCacheTestDriver.

File: server/hotrod/pom.xml [hunk 2]
```diff
@@ -41,7 +41,6 @@
       <dependency>
          <groupId>${project.groupId}</groupId>
          <artifactId>infinispan-tasks</artifactId>
-         <optional>true</optional>
```
Comment (wburns) [inline server/hotrod/pom.xml]: Hrmm this seems unrelated.

File: core/src/test/java/org/infinispan/security/SecureCacheTestDriver.java [hunk 3]
```diff
@@ -676,4 +659,7 @@ public void testAddFilteredListener_Object_CacheEventFilter_CacheEventConverter_
       cache.addFilteredListener(listener, keyValueFilter, converter, Collections.emptySet());
    }
 
+   @TestCachePermission(AuthorizationPermission.NONE)
+   public void testWithSubject_Subject(SecureCache<String, String> cache) {}
```
Comment (wburns) [inline core/src/test/java/org/infinispan/security/SecureCacheTestDriver.java]: Did you want to put some code here?

Comment (tristantarrant) [inline core/src/test/java/org/infinispan/security/SecureCacheTestDriver.java]: No, this is fine

**Comment by slaskawi** [comment]
Those 2 failures look suspicious:
```
org.infinispan.server.test.configs.ExampleConfigsIT.testSSLHotRodConfig
org.infinispan.integrationtests.spring.boot.session.remote.RemoteSpringSessionTest.org.infinispan.integrationtests.spring.boot.session.remote.RemoteSpringSessionTest
```

Let me retrigger CI, maybe it happened by an accident.

**Comment by ryanemerson** [comment]
`ExampleConfigsIT.testSSLHotRodConfig` seems to be failing for several builds, so I don't think that failure is related. However, I have tested `RemoteSpringSessionTest` locally and it fails on every run with a `IndexOutOfBoundsException`, so changes are required

**Comment by tristantarrant** [comment]
I fixed the failing test.

**Comment by ryanemerson** [comment]
Integrated, thanks @tristantarrant
